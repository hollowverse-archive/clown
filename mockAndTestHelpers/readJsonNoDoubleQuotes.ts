import { fs } from '@forabi/memfs';

/* Jest, when doing snapshots, wraps strings in double quotes. Since JSON keys and string values
are also wrapped in double quotes, Jest finds itself needing to escape all the double quotes in a
JSON, so the JSON becomes cluttered with \\" \\" \\" \\", making it unreadable. The function
below substitutes double quotes for single quotes when reading JSON content from the mock disk,
so there wouldn't be double-quote escaping issues. */
export function readJsonNoDoubleQuotes(filePathInDisk: string) {
  // tslint:disable-next-line:quotemark
  return (fs.readFileSync(filePathInDisk, 'utf8') as string).replace(/"/g, "'");
}
