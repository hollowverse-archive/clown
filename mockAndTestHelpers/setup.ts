jest.mock('fs');
process.exit = jest.fn() as any;
jest.mock('../src/printErrors');
