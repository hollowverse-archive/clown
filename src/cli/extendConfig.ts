import path from 'path';
import { writeFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import fs from 'fs-extra';

export async function extendConfig(cwd: string) {
  const fileContents = await computeFileContents(cwd);
  const iterableFileContents = _.entries(fileContents);

  await bluebird.each(
    iterableFileContents,
    async ([destinationPath, fileContent]) => {
      await fs.mkdirp(path.dirname(destinationPath));

      return writeFile(destinationPath, fileContent);
    },
  );
}
