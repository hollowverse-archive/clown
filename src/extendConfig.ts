import path from 'path';
import { writeFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import fs from 'fs-extra';
import { isJsonEqual } from './isJsonEqual';

export async function extendConfig(cwd: string) {
  const fileContents = await computeFileContents(cwd);
  const iterableFileContents = _.entries(fileContents);

  await bluebird.each(
    iterableFileContents,
    async ([destinationPath, fileContent]) => {
      if (
        fileContent.destinationContent &&
        fileContent.type === 'json' &&
        isJsonEqual(fileContent.destinationContent, fileContent.computedContent)
      ) {
        return;
      }

      await fs.mkdirp(path.dirname(destinationPath));

      return writeFile(destinationPath, fileContent.computedContent);
    },
  );
}
