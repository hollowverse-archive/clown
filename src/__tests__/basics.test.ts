import { extendConfig } from '../extendConfig';
import { readJsonNoDoubleQuotes } from '../../mockAndTestHelpers/readJsonNoDoubleQuotes';
import { vol } from '@forabi/memfs';

describe('Basics', () => {
  beforeEach(() => {
    vol.reset();
  });

  it('use case 1', async () => {
    const files = {
      '/clown.json': `{
        "extensions": ["/clownExtensionA"]
      }`,
      '/clownExtensionA/package.json': `{
        "foo": "bar"
      }`,
      '/package.json': `{
        "name": "Stuff"
      }`,
    };

    vol.fromJSON(files);

    await extendConfig('/');

    expect(readJsonNoDoubleQuotes('/package.json')).toMatchSnapshot();
  });

  it('does not rewrite a JSON file content if the JSON values are equal', async () => {
    const files = {
      '/clown.json': `{
        "extensions": ["/clownExtensionA"]
      }`,
      '/clownExtensionA/.eslintrc.json': `{
        "extends": [
          "./base/.eslintrc.json"
        ]
      }`,
      '/.eslintrc.json': `{
        "extends": ["./base/.eslintrc.json"]
      }`,
    };

    vol.fromJSON(files);

    await extendConfig('/');

    expect(readJsonNoDoubleQuotes('/.eslintrc.json')).toMatchSnapshot();
  });
});
