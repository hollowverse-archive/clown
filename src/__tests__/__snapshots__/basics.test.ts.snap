// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basics does not rewrite a JSON file content if the JSON values are equal 1`] = `
"{
        'extends': ['./base/.eslintrc.json']
      }"
`;

exports[`Basics use case 1 1`] = `
"{
        'name': 'Stuff',
        'foo': 'bar'
}
"
`;
